### PR TITLE
VertexAI feature

### PR DESCRIPTION
**Pull request by BallenAlain**
#30 

- Log exceptions
- Model checks

School accounts work, we just need a credit card. I can use my own, but if there's an alternative option, please let me know.

### CODE REVIEW

File: llmproxy/models/vertexai.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            parameters = {
+                "temperature": self.temperature,  # Temperature controls the degree of randomness in token selection.
+                "max_output_tokens": 256,  # Token limit determines the maximum amount of text output.
+                "top_p": 0.8,  # Tokens are selected from most probable to least until the sum of their probabilities equals the top_p value.
+                "top_k": 40,  # A top_k of 1 means the selected token is the most probable among all tokens.
+            }
+
+            chat_model = TextGenerationModel.from_pretrained(self.model)
+            response = chat_model.predict(self.prompt)
+            output = f"Response from Model: {response.text}"
```
Comment (laxitshahi) [inline llmproxy/models/vertexai.py:43]: Return just the response in text form for now

File: llmproxy/models/vertexai.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        prompt: str = "",
+        temperature: float = 0,
+        model: VertexAIModel = VertexAIModel.PALM_TEXT,
+        project_id: str | None = "",
+        location: str = "",
+    ) -> None:
+        self.prompt = prompt
+        self.temperature = temperature
+        self.model = model
+        aiplatform.init(project=project_id, location=location)
```
Comment (laxitshahi) [inline llmproxy/models/vertexai.py:24]: If passing in a location that DNE, then you should have a check/default value for that

File: llmproxy/models/vertexai.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    PALM_CHAT = "chat-bison"
+
+class VertexAI(BaseChatbot):
+    def __init__(
+        self,
+        prompt: str = "",
+        temperature: float = 0,
+        model: VertexAIModel = VertexAIModel.PALM_TEXT,
+        project_id: str | None = "",
+        location: str = "",
```
Comment (laxitshahi) [inline llmproxy/models/vertexai.py:19]: If it's empty location, what happens?

File: llmproxy/models/vertexai.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+class VertexAIModel(str, BaseEnum):
+    PALM_TEXT = "text-bison@001"
+    PALM_CHAT = "chat-bison"
+
+class VertexAI(BaseChatbot):
+    def __init__(
+        self,
+        prompt: str = "",
+        temperature: float = 0,
+        model: VertexAIModel = VertexAIModel.PALM_TEXT,
```
Comment (laxitshahi) [inline llmproxy/models/vertexai.py:17]: There is a small bug here, it should be VertexAIModel.PALM_TEXT.value

**Review by laxitshahi — CHANGES_REQUESTED** [review]
Make sure you also add:
- Unit tests
- ~google_credentials to gitignore (or whatever you end up naming the file)~

File: llmproxy/models/vertexai.py (hunk 3 above)
Comment (BallenAlain) [reply llmproxy/models/vertexai.py]: turns out it functions as normal

File: llmproxy/models/vertexai.py (hunk 2 above)
Comment (BallenAlain) [reply llmproxy/models/vertexai.py]: default value will be an empty string

File: llmproxy/tests/unit_tests/models/test_vertexai.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+load_dotenv(".env.test")
+project_id = os.getenv("GOOGLE_PROJECT_ID")
+
+def test_invalid_project_id() -> None:
+    vertexai = VertexAI(project_id="invalid id")
+    response = vertexai.get_completion()
+    assert "Permission denied" in response.message
+
+def test_invalid_location() -> None:
```
Comment (laxitshahi) [inline llmproxy/tests/unit_tests/models/test_vertexai.py:14]: Does this api call cost money even though it fails? If so we need to ensure that it is mocked or the error is handled before the api call is made

File: llmproxy/tests/unit_tests/models/test_vertexai.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import os
+
+from llmproxy.models.vertexai import VertexAI
+from dotenv import load_dotenv
+
+load_dotenv(".env.test")
+project_id = os.getenv("GOOGLE_PROJECT_ID")
+
+def test_invalid_project_id() -> None:
```
Comment (laxitshahi) [inline llmproxy/tests/unit_tests/models/test_vertexai.py:9]: This function depends on a valid application_default_credentials.json, this dependency needs to be stated or seperated

File: llmproxy/tests/unit_tests/models/test_vertexai.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    vertexai = VertexAI(project_id="invalid id")
+    response = vertexai.get_completion()
+    assert "Permission denied" in response.message
+
+def test_invalid_location() -> None:
+    vertexai = VertexAI(project_id=project_id, location="test")
+    response = vertexai.get_completion()
+    assert "Unsupported region" in response.message
+
+def test_unsupported_model() -> None:
```
Comment (laxitshahi) [inline llmproxy/tests/unit_tests/models/test_vertexai.py:19]: Would be best if you did Arrange, Act, Assert pattern explicitly: Check Mistral unit tests as an example

File: llmproxy/tests/unit_tests/models/test_vertexai.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    vertexai = VertexAI(project_id=project_id,model="test")
+    response = vertexai.get_completion()
+    assert "Model not supported" in response.message
+
```
Comment (laxitshahi) [inline llmproxy/tests/unit_tests/models/test_vertexai.py:23]: Also getting 21 warnings about deprecation, we need to make sure this won't affect the codebase in the future

**Review by laxitshahi — CHANGES_REQUESTED** [review]

File: llmproxy/tests/unit_tests/models/test_vertexai.py (hunk 6 above)
Comment (BallenAlain) [reply llmproxy/tests/unit_tests/models/test_vertexai.py:9]: unit tests assume we have a valid credentials json file

File: llmproxy/tests/unit_tests/models/test_vertexai.py (hunk 7 above)
Comment (BallenAlain) [reply llmproxy/tests/unit_tests/models/test_vertexai.py:29]: added

File: llmproxy/tests/unit_tests/models/test_vertexai.py (hunk 8 above)
Comment (BallenAlain) [reply llmproxy/tests/unit_tests/models/test_vertexai.py:38]: looking in to it

**Pull request merged**